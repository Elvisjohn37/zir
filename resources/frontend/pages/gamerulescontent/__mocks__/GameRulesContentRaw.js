import { componentModuleMock } from 'frontend/tests/helpers';
module.exports = componentModuleMock(jest.createMockFromModule('../GameRulesContentRaw'), 'GameRulesContentRaw');
