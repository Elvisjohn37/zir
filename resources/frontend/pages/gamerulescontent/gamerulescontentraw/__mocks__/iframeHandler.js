
let isSuccess;

let loadGoogleDocsIframe = jest.fn((iframe, url, callback) => {
    if(isSuccess) {
        callback.success && callback.success();
    } else {
        callback.error
    }
});

let resetLoadGoogleDocsIframe = () => {
    isSuccess = true;
}

let setError = () => {
    isSuccess = false;
}

resetLoadGoogleDocsIframe();

export { loadGoogleDocsIframe, resetLoadGoogleDocsIframe, setError };