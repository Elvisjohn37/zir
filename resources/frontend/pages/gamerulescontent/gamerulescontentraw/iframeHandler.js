import { googleDocs } from 'frontend/ajax/google';

export function loadGoogleDocsIframe(iframe, url, callback) {
	googleDocs(
		url,
		(response) => {
			iframe.contentWindow.document.getElementsByTagName('html')[0].innerHTML = response.data;
			iframe.contentWindow.document.getElementsByTagName('head')[0].innerHTML = response.data;

			let srcollHeight = iframe.contentWindow.document.body.scrollHeight;
			iframe.style.height = `${srcollHeight}px`;
			iframe.contentWindow.document.body.style.background = 'transparent';

			callback.success && callback.success();
		},
		callback.error
	);
}
