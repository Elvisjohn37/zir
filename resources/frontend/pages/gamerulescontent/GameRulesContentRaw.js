import React, { useState } from 'react';
import { useParams, useLocation, NavLink } from 'react-router-dom';
import Container from '@mui/material/Container';
import IconButton from '@mui/material/IconButton';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';
import styles from './GameRulesContentRaw.module';
import SeamlessIframe from 'frontend/components/SeamlessIframe';
import { loadGoogleDocsIframe } from './gamerulescontentraw/iframeHandler';
import classnames from 'classnames';
import lodash from 'lodash';
import GameRulesErrorMessage from 'frontend/components/GameRulesErrorMessage';
import GameRulesLoader from 'frontend/components/GameRulesLoader';

export default function GameRulesContentRaw({ gameRulesState }) {
	let [isContentLoading, setIsContentLoading] = useState(true);
	let [isContentError, setIsContentError] = useState(false);

	let { gameName } = useParams();
	let gameRuleDetails = gameRulesState.gameRulesList.find((item) => lodash.kebabCase(item.gameName) == gameName);

	let location = useLocation();
	let splitPathName = location.pathname.split('/');
	splitPathName.pop();
	let rootPath = splitPathName.join('/');

	let iframeRef = React.useRef();

	return (
		<Container className={styles.gameRulesContentRaw}>
			<div className={styles.backButtonContainer}>
				<IconButton component={NavLink} to={rootPath}>
					<ArrowBackIcon />
				</IconButton>
			</div>
			<div className={styles.content}>
				{gameRulesState.isLoading ? (
					<GameRulesLoader />
				) : gameRulesState.isError || gameRuleDetails === undefined || isContentError ? (
					<GameRulesErrorMessage />
				) : (
					<>
						{isContentLoading && <GameRulesLoader />}
						<SeamlessIframe
							scrolling="no"
							src={'/frame'}
							iframeHolderClassName={classnames([styles.iframeHolder, isContentLoading && styles.isContentLoading])}
							iframeClassName={styles.iframe}
							ref={iframeRef}
							width={640}
							onLoad={(e) => {
								loadGoogleDocsIframe(e.target, gameRuleDetails.url, {
									success: () => {
										iframeRef.current.resize();
										setIsContentLoading(false);
									},
									error: () => {
										setIsContentError(true);
									},
								});
							}}
						/>
					</>
				)}
			</div>
		</Container>
	);
}
