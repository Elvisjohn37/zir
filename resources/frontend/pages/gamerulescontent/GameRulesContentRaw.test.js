import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import GameRulesContentRaw from './GameRulesContentRaw';
import * as gameRulesHelper from 'frontend/tests/gameRulesHelper';
import * as reactRouterDom from 'react-router-dom';
import * as iframeHandler from './gamerulescontentraw/iframeHandler';
import lodash from 'lodash';

jest.mock('react-router-dom');
jest.mock('@mui/material/Container');
jest.mock('@mui/material/IconButton');
jest.mock('@mui/icons-material/ArrowBack');
jest.mock('frontend/components/SeamlessIframe');
jest.mock('./gamerulescontentraw/iframeHandler');
jest.mock('classnames');
jest.mock('lodash');
jest.mock('frontend/components/GameRulesErrorMessage');
jest.mock('frontend/components/GameRulesLoader');

describe('<GameRulesContentRaw />', () => {
	beforeEach(() => {
		gameRulesHelper.resetGameRulesState();
		reactRouterDom.resetUseParams();
	});

	it('should match snapshot', () => {
		reactRouterDom.setUseParams({
			gameName: 'dice-wars-mobile',
		});
		let { asFragment } = render(<GameRulesContentRaw gameRulesState={gameRulesHelper.getGameRulesState()} />);
		expect(asFragment()).toMatchSnapshot();
	});

	it('should display loader when game rules content is loading', () => {
		reactRouterDom.setUseParams({
			gameName: 'dice-wars-mobile',
		});

		gameRulesHelper.setGameRulesState({
			isLoading: true,
		});

		let { queryByTestId, getByTestId, rerender } = render(
			<GameRulesContentRaw gameRulesState={gameRulesHelper.getGameRulesState()} />
		);

		expect(getByTestId('GameRulesLoader')).toBeInTheDocument();

		gameRulesHelper.setGameRulesState({
			isLoading: false,
		});

		rerender(<GameRulesContentRaw gameRulesState={gameRulesHelper.getGameRulesState()} />);

		expect(getByTestId('GameRulesLoader')).toBeInTheDocument();

		fireEvent.load(getByTestId('SeamlessIframe'));

		expect(queryByTestId('GameRulesLoader')).not.toBeInTheDocument();
	});

	it('should display game rules after loading', () => {
		jest.spyOn(iframeHandler, 'loadGoogleDocsIframe');
		reactRouterDom.setUseParams({
			gameName: 'yin-yang-treasure-mobile',
		});

		let gameDetails = gameRulesHelper
			.getGameRulesState()
			.gameRulesList.find((game) => lodash.kebabCase(game.gameName) == 'dice-wars-mobile');
		let { getByTestId } = render(<GameRulesContentRaw gameRulesState={gameRulesHelper.getGameRulesState()} />);
		let iframe = getByTestId('SeamlessIframe');
		fireEvent.load(iframe);

		expect(iframe.getAttribute('iframeholderclassname').includes('isContentLoading')).toBeFalsy();
		expect(iframeHandler.loadGoogleDocsIframe).toHaveBeenCalledWith(iframe, gameDetails.url, {
			success: expect.any(Function),
			error: expect.any(Function),
		});
	});

	it('should display error message when game rules data loading failed', () => {
		reactRouterDom.setUseParams({
			gameName: 'yin-yang-treasure-mobile',
		});

		let { getByTestId, queryByTestId, rerender } = render(
			<GameRulesContentRaw gameRulesState={gameRulesHelper.getGameRulesState()} />
		);

		expect(queryByTestId('GameRulesErrorMessage')).not.toBeInTheDocument();

		gameRulesHelper.setGameRulesState({
			isError: true,
		});

		rerender(<GameRulesContentRaw gameRulesState={gameRulesHelper.getGameRulesState()} />);

		expect(getByTestId('GameRulesErrorMessage')).toBeInTheDocument();
	});
});
