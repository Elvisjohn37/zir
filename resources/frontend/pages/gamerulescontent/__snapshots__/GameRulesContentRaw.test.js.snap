// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameRulesContentRaw /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="Container"
  >
    <div
      class="backButtonContainer"
    >
      <button
        data-testid="IconButton"
        to="www.zirconloc.com"
      >
        <div
          data-testid="ArrowBack"
        />
        <div>
          {}
        </div>
      </button>
    </div>
    <div
      class="content"
    >
      <div
        data-testid="GameRulesLoader"
      />
      <iframe
        data-testid="SeamlessIframe"
        iframeclassname="iframe"
        iframeholderclassname="iframeHolder isContentLoading"
        scrolling="no"
        src="/frame"
        width="640"
      />
    </div>
  </div>
</DocumentFragment>
`;
